produce_set.go

```diff
@@ -90,11 +90,29 @@ func (ps *produceSet) buildRequest() *ProduceRequest {
 					Logger.Println(err) // if this happens, it's basically our fault.
 					panic(err)
 				}
-				req.AddMessage(topic, partition, &Message{
+				compMsg := &Message{
 					Codec: ps.parent.conf.Producer.Compression,
 					Key:   nil,
 					Value: payload,
-				})
+				}
+				if ps.parent.conf.Version.IsAtLeast(V0_10_0_0) {
+					// Compressed messages must use a protocol version
+					// that is newer than the inner messages version.
+					// Due to a lack of better timestamp notation copy the oldest
+					// (earliest) timestamp to message.
+					for _, msgBlock := range set.setToSend.Messages {
+						msg := msgBlock.Msg
+						if msg.Version > compMsg.Version {
+							compMsg.Version = msg.Version
```

Comment:
Hmm, I was thinking about better ways to accomplish this and I wonder now if the code in `produceSet#add` is even correct; is it safe to mix v0 and v1 messages in a single compressed batch which we'll do right now if some timestamps are 0? Or should we change to just always using v1 if the version supports it, and let the timestamps be 0 if necessary?

edit: actually this issue goes away if we provide a timestamp of `time.Now()` by default if you leave the timestamp as 0, which we should probably do for convenience anyway